crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -93,22 +116,278 @@ pub struct TableMetadata {
+                snapshot_id: -1,
+                retention: SnapshotRetention::Branch {
+                    min_snapshots_to_keep: None,
+                    max_snapshot_age_ms: None,
+                    max_ref_age_ms: None,
+                },
+            },
+        )])
+    }
+
+    /// Add or replace a snapshot_reference
+    /// branch : branch id of the snapshot
+    /// snapshot_ref : SnapshotReference to add or update
+    fn with_ref(&mut self, branch: String, snapshot_ref: SnapshotReference) -> &mut Self {
+        if branch == "main" {
+            self.current_snapshot_id = Some(Some(snapshot_ref.snapshot_id));
+            if let Some(vec) = self.snapshot_log.as_mut() {
+                vec.push(SnapshotLog {
+                    snapshot_id: snapshot_ref.snapshot_id,
+                    timestamp_ms: self.last_updated_ms.unwrap_or(Self::current_time_ms()),
+                })
+            } else {
+                self.snapshot_log = Some(vec![SnapshotLog {
+                    snapshot_id: snapshot_ref.snapshot_id,
+                    timestamp_ms: self.last_updated_ms.unwrap_or(Self::current_time_ms()),
+                }])
+            }
+        }
+        if let Some(map) = self.refs.as_mut() {
+            map.insert(branch, snapshot_ref);
+        } else {
+            self.refs = Some(HashMap::from([(branch, snapshot_ref)]));
+        }
+        self
+    }
+
+    /// Initialize a TableMetadata with a TableCreation struct
+    /// the Schema, sortOrder and PartitionSpec will be set as current
+    pub fn from_table_creation(&mut self, tc: TableCreation) -> &mut Self {
+        self.with_location(tc.location)
+            .with_properties(tc.properties)
+            .with_default_sort_order(tc.sort_order)
+            .with_current_schema(tc.schema);
+
+        if let Some(partition_spec) = tc.partition_spec {
+            self.with_default_partition_spec(partition_spec);
+        }
+        self
+    }
+
+    /// Add or replace a schema
+    /// schema : Schema to be added or replaced
+    pub fn with_schema(&mut self, schema: Schema) -> &mut Self {
+        if let Some(map) = self.schemas.as_mut() {
+            map.insert(schema.schema_id(), Arc::new(schema));
+        } else {
+            self.schemas = Some(HashMap::from([(schema.schema_id(), Arc::new(schema))]));
+        }
+        self
+    }
+
+    /// Add or replace a schema and set current schema to this one
+    /// schema : Schema to be added or replaced
+    pub fn with_current_schema(&mut self, schema: Schema) -> &mut Self {
+        self.current_schema_id = Some(schema.schema_id());
+        self.last_column_id = Some(schema.highest_field_id());
+        self.with_schema(schema)
+    }
+
+    /// Add or replace a partition_spec and update the last_partition_id accordingly
+    /// partition_spec : PartitionSpec to be added or replaced
+    pub fn with_partition_spec(&mut self, partition_spec: PartitionSpec) -> &mut Self {
```

Comment:
```suggestion
    pub fn add_partition_spec(&mut self, partition_spec: PartitionSpec) -> &mut Self {
```

1. There are same problems as scheme id.
2. We need to validate spec with schema: 
  * If `source_id` exists in schema
  * If the field type supports such transformation.